### PR TITLE
Global settings

### PR DESCRIPTION
**Pull request by ackagel**
### **PR Purpose**

Adds global settings for data column-name standardization.  Closes issue #187 .

### **Changes**

Added `settings.py` into `ark` folder.  `settings.py` contains several global constants which contain common cell data column names.  In place of these string-literal names, settings have been imported and used.

Channels are now assumed to be between `PRE_CHANNEL_COL` and `POST_CHANNEL_COL` which are aliases to other constants in `settings.py`.

The example expression matrix csv was edited to follow this naming scheme.

### **Remaining Issues**
If we change standards/names from within `settings.py`, everything code-wise should still work fine.  However, the example expression matrix csv will need to be changed to follow this convention.  A possible work around, would be to store a different version of the expression matrix csv in a different file, and dynamically reorder/rename its columns to match `settings.py`.  Note, this is probs overkill if we're not going to change standards/names several times, but not too hard to implement.

### CODE REVIEW

**Comment by ackagel** [comment]
by no means is this ready to go; just need some feedback on the current approach

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Yes, I think this is the right approach. A couple things

1. We should organize the settings.py with comments, similar to how the .gitignore is set up, to indicate what these variables are doing

2. Users won't be able to change settings.py. The goal is that rather than hardcoding in the defaults, the defaults get taken from settings. So we should still have these be modifiable with a different argument to the function, but in the absence of user-intervention, these are the defaults that are used, rather than whatever happens to be defined in the function def

3. I think that defining an exhaustive list of pre-channels and post-channels columns won't work, since people will be adding all kinds of random columns onto this CSV, and we want it to still work. 

It's slightly more work, but I think it's worth defining the first and last channel isntead, and then indexing using that value directly. See here: https://github.com/angelolab/ark-analysis/blob/1a229990f19141be1eaeda5e7d56566efc4d5d51/ark/utils/segmentation_utils.py#L115

Open to other suggestions as well

**Review by alex-l-kong — COMMENTED** [review]
I think Noah covered most everything. Good start!

**Review by saeedseyyedi — COMMENTED** [review]
Looks good so far + agree on Noah's comments.

**Comment by ackagel** [comment]
Not ready to merge, but definitely time for a quick review here.

There are areas where instead of exhaustively excluding column names (see spatial analysis notebook) we could be just be excluding channel names, now that the channel locations within cell tables are 'standardized'.  If you know of and/or can find places where this standardization can be applied, lmk.

Also, `example_expression_matrix.csv` might need to have its columns changed (by order and possibly name) to conform with the new standard.  Furthermore, how leniently/strictly should we enforce this standard for names?  In particular, there are a few hardcoded, non-default settings for label column names, ex:

https://github.com/angelolab/ark-analysis/blob/e21b33413040217819c6ebf83f9ceb7fa8b2211a/ark/utils/spatial_analysis_utils.py#L164-L165

The new names conflict with the old names in past versions of generated expression matrices.  If we remove these, and want backwards compatibility, we'll have to pass down these values thru several functions, which is fine if we're going to kwarg-it-up, but just want to do a vibe check on that before I implement anything.

Let me know y'all's thoughts one everything above.

**Comment by ngreenwald** [comment]
Agreed. We should update the example data to have our enforced column
names. The vast majority of people using this pipeline haven't generated
their data yet, so it's okay to make breaking changes. People can update
their column names accordingly.

For now, let's default to not having kwargs. I'm sure there are some
functions where we definitely will want to add that in, but it's hard to
say now which ones they are. I think for a first pass it can be none, and
then we decide which it's worth it to add in.

**Review by alex-l-kong — COMMENTED** [review]
Looks good from what I saw! Definitely helpful not to hard code column names and stuff like that.

File: ark/analysis/spatial_analysis.py
Comment (ngreenwald) [inline ark/analysis/spatial_analysis.py:33]: I think we only need to explain the defaults if the value triggers different behavior. For example, if None then takes all images. Otherwise, we're just repeating what's already there.

File: ark/analysis/spatial_analysis.py
Comment (ngreenwald) [inline ark/analysis/spatial_analysis.py:54]: I think this will fail for None. Either way, we should add a test checking for when it's None

File: ark/analysis/spatial_analysis.py
Comment (ngreenwald) [inline ark/analysis/spatial_analysis.py:135]: Same thing. I don't think there's any benefit to explaining what the default is when the default is in the arg list

File: ark/analysis/spatial_analysis_test.py
Comment (ngreenwald) [inline ark/analysis/spatial_analysis_test.py:24]: I don't think these need to be global settings if their only purpose is to properly name the test data. We're never going to change the names that come directly from regionprops

File: ark/settings.py
Comment (ngreenwald) [inline ark/settings.py:10]: We'll probably need to wait until next Friday's meeting to merge this in. I'm sure Mike will have opinions about what the default names should be

File: ark/utils/spatial_analysis_utils.py
Comment (ngreenwald) [inline ark/utils/spatial_analysis_utils.py:169]: What are these used for?

File: ark/utils/spatial_analysis_utils_test.py
Comment (ngreenwald) [inline ark/utils/spatial_analysis_utils_test.py:154]: Can we use PRE_CHANNEL and POST_CHANNEL here?

**Review by ngreenwald — APPROVED** [review]
Great progress! Some more comments below

File: ark/utils/spatial_analysis_utils.py
Comment (ackagel) [reply ark/utils/spatial_analysis_utils.py:169]: aren't they used in the `get_pos_cell_labels_` functions?

Comment (ngreenwald) [reply ark/utils/spatial_analysis_utils.py:169]: Ah, I didn't see any diff adding them in, but that's because they were already defined in the function args. Can you set the defaults there?

File: ark/utils/spatial_analysis_utils_test.py
Comment (ackagel) [reply ark/utils/spatial_analysis_utils_test.py:154]: yep.  missed that one.

**Comment by ackagel** [comment]
don't want to merge it yet (wanna check on names and see if travis actually passes it once we resolve the urllib3 issue), but just want some feedback in the meantime since its been awhile.

**Review by alex-l-kong — COMMENTED** [review]
Looking good so far, must've been a pain to handle all those merge conflicts.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
This is great, going to be really helpful once the rest of the lab starts using these and everyone has different names for everything

**Comment by alex-l-kong** [comment]
@ackagel for the time being, to address the Travis issues, you can simply delete the line in `.travis.yml` that runs `python setup.py install`.

**Review by ngreenwald — APPROVED** [review]

**Pull request merged**
